README.md
# Microsoft Word
  
Module to work with Microsoft Word  

*Read this in other languages: [English](README.md), [Portugues](README.pr.md), [Español](README.es.md).*

## How to install this module
  
__Download__ and __install__ the content in 'modules' folder in Rocketbot path  



## Overview


1. New Document  
Create a new word document

2. Open Document  
Open a Word document.

3. Read Document  
Extract text from a Word document

4. Write in Document  
Write in a Word document.

5. Copy and paste text  
Copy and paste text between ranges in a Word document and paste it in another document

6. Copy text  
Copy text to clipboard between ranges in a Word document

7. Paste text  
Paste text from clipboard in a Word document

8. Count characters  
Count characters in a specific paragraph

9. Add table  
Add table in a Word document.

10. Read Tables  
Extract data from the Tables in the document

11. Edit table  
Edit table from a Word document.

12. Add Page  
Add a new page to the document

13. Add Picture  
Add an image to the document.

14. Convert to PDF  
Convert Word document to PDF.

15. Locate Text in Paragraph  
Locate in which paragraph there is an indicated text.

16. Count Paragraphs  
Count the number of paragraphs in the document. Includes table fields.

17. Replace text in paragraph  
Replace the text of a paragraph.

18. Delete paragraph  
Delete a paragraph from the document. If tables are included, the Find Text in Paragraph command should be used to locate the paragraph to be deleted.

19. Add text at the end of bookmark  
Add text at the end of bookmark.

20. Save document  
Extract text from file.

21. Close Document  
Close the document that is running  




----
### OS

- windows

### Dependencies
- [**pywin32**](https://pypi.org/project/pywin32/)
### License
  
![MIT](https://camo.githubusercontent.com/107590fac8cbd65071396bb4d04040f76cde5bde/687474703a2f2f696d672e736869656c64732e696f2f3a6c6963656e73652d6d69742d626c75652e7376673f7374796c653d666c61742d737175617265)  
[MIT](http://opensource.org/licenses/mit-license.ph)

CHANGES.txt
Thu Jul 21 01:32:22 2022  Merge branch qa into branch-nico

html/message.js

var data_

function dataHandler(e){

    let lineType = document.getElementById('lineType')
    let lineSize = document.getElementById('lineSize')

    console.log('Parent received message!: ', e.data)
    if (e.data && e.data.lineSize && e.data.lineType ) {

            dataLineSize = e.data.lineSize
            dataLineType = e.data.lineType

    } 

    // if (e.data.lineSize == ''){
    //     dataLineSize = 'None'
    // }

    // if (e.data.lineType == ''){
    //     dataLineType = '100_point'
    // }
    
    lineType.value = dataLineType
    lineSize.value = dataLineSize

    return
}



var message = {
    type: 'iframe',
    commands: {}
}

var SendMessage = function() {
    parent.postMessage(message, '*')
}



var eventMethod = window.addEventListener ? "addEventListener" : "attachEvent"
var eventer = window[eventMethod]
var messageEvent = eventMethod == "attachEvent" ? "onmessage" : "message"

eventer(messageEvent, dataHandler)






html/options.js
let lineType = document.getElementById('lineType')
let lineSize = document.getElementById('lineSize')
let btnSelect = document.getElementById('btnSelect')



let styles = ["Dash dot", 
               "Dash and two dots", 
               "Dash dot stroked", 
               "Dash with large gap", 
               "Dash with small gap", 
               "Dotted", 
               "Double", 
               "Double wavy", 
               "3D Embossed look", 
               "3D Engraved look", 
               "Inset", 
               "None", 
               "Outset", 
               "Single", 
               "Single wavy", 
               "Thick thin large gap", 
               "Thick thin med gap", 
               "Thick thin small gap", 
               "Thin thick large gap", 
               "Thin thick med gap", 
               "Thin thick small gap", 
               "Thin thick thin large gap", 
               "Thin thick thin med gap", 
               "Thin thick thin small gap", 
               "Triple"]

let sizes = ["1/4 point", 
               "1/2 point", 
               "3/4 point", 
               "1 point", 
               "1 1/2 points", 
               "2 1/4 points", 
               "3 points", 
               "4 1/2 points", 
               "6 points"]

let valuesSizes = ["25_point",
                   "50_point", 
                   "75_point", 
                   "100_point", 
                   "150_points", 
                   "225_points", 
                   "300_points", 
                   "450_points", 
                   "600_points"]




function showSizes(array, arrayValues)  {

    while (lineSize.length > 0) {
        lineSize.remove(0);
    }

    for(item in array){
        let option = document.createElement('option')
        option.value = arrayValues[item]
        option.text = array[item]
        lineSize.appendChild(option)
    } 
}

lineType.addEventListener('change', function(){
    let valor = lineType.value
    let cut = []
    let cutSizes = []

    switch (valor) {
        case 'thin_thick_thin_med_gap':
            cut = sizes.slice(0, 8)
            cutSizes = valuesSizes.slice(0, 8)
            showSizes(cut, cutSizes, lineSize)
        break;
        case 'single_wavy':
            cut = ["3/4 point", "1 1/2 points"]
            cutSizes = ["75_point", "150_points"]
            showSizes(cut, cutSizes, lineSize)

        break;
        case 'double_wavy':
            cut = ["3/4 point"]
            cutSizes = ["75_point"]
            showSizes(cut, cutSizes, lineSize)
        break;
        case 'dash_dot_stroked':
            cut = ["3 points"]
            cutSizes = ["300_points"]
            showSizes(cut, cutSizes, lineSize)
        break;
        case 'double':
            cut = sizes.slice(0, 7)
            cutSizes = valuesSizes.slice(0, 7)
            showSizes(cut, cutSizes, lineSize)
        break;
        case 'triple':
            cut = sizes.slice(0, 7)
            cutSizes = valuesSizes.slice(0, 7)
            showSizes(cut, cutSizes, lineSize)
        break;
        default:
            showSizes(sizes, valuesSizes, lineSize)
            break;
    }

    message.commands["lineType"] = lineType.value
    message.commands["lineSize"] = lineSize.value
    SendMessage()
})

lineSize.addEventListener('change', function(){
    message.commands["lineType"] = lineType.value
    message.commands["lineSize"] = lineSize.value
    SendMessage()
})

// btnSelect.addEventListener('click', function(){
//     message.commands["lineType"] = lineType.value
//     message.commands["lineSize"] = lineSize.value
//     SendMessage()
// })



__init__.py
# coding: utf-8
"""
Base para desarrollo de modulos externos.
Para obtener el modulo/Funcion que se esta llamando:
     GetParams("module")

Para obtener las variables enviadas desde formulario/comando Rocketbot:
    var = GetParams(variable)
    Las "variable" se define en forms del archivo package.json

Para modificar la variable de Rocketbot:
    SetVar(Variable_Rocketbot, "dato")

Para obtener una variable de Rocketbot:
    var = GetVar(Variable_Rocketbot)

Para obtener la Opcion seleccionada:
    opcion = GetParams("option")


Para instalar librerias se debe ingresar por terminal a la carpeta "libs"

    pip install <package> -t .

"""
import os
import sys
import re

base_path = tmp_global_obj["basepath"]
cur_path = base_path + 'modules' + os.sep + 'MicrosoftWord' + os.sep + 'libs' + os.sep
sys.path.append(cur_path)

# Import local libraries
import win32com.client

global mod_microsoft_word
SESSION_DEFAULT = "default"
try:
    if not mod_microsoft_word: #type:ignore
        mod_microsoft_word = {SESSION_DEFAULT: {}}
except NameError:
    mod_microsoft_word = {SESSION_DEFAULT: {}}


module = GetParams("module")
session = GetParams("session")
if not session:
    session = SESSION_DEFAULT
    

try:
    ms_word = mod_microsoft_word[session].get("app", None)
    word_document = mod_microsoft_word[session].get("doc", None)
except:
    pass

def alignments(WdParagraphAlignment):
    return ["Left", "Center", "Rigth", "Justify"][WdParagraphAlignment]


WdBuiltinStyle = {
    "paragraph": -1,
    "heading1": -2,
    "heading2": -3,
    "heading3": -4,
    "heading4": -5,
    "heading5": -6,
    "heading6": -7,
    "heading7": -8,
    "heading8": -9,
    "heading9": -10,
    "caption": -35,
    "bullet1": -49,
    "number1": -50,
    "bullet2": -55,
    "bullet3": -56,
    "bullet4": -57,
    "bullet5": -58,
    "number2": -59,
    "number3": -60,
    "number4": -61,
    "number5": -62,
    "title": -63,
    "subtitle": -75,
    "colorful_grid": -172,
    "colorful_list": -171,
    "colorful_shading": -170,
    "dark_list": -169,
    "light_grid": -161,
    "light_grid_accent_1": -175,
    "light_list": -160,
    "light_list_accent_1": -174,
    "light_shading": -159,
    "light_shading_accent_1": -173,
    "medium_grid_1": -166,
    "medium_grid_2": -167,
    "medium_grid_3": -168,
    "medium_list_1": -164,
    "medium_list_accent_1": -178,
    "medium_list_2": -165,
    "medium_shading_1": -162,
    "medium_shading_1_accent_1": -176,
    "medium_shading_2": -163,
    "medium_shading_2_accent_1": -177,
    "quote": -181,
    "intense_quote": -182,
    "book": -265
}

WdLineWidth = {
    "25_point": 2,
    "50_point": 4,
    "75_point": 6,
    "100_point": 8,
    "150_points": 12,
    "225_points": 18,
    "300_points": 24,
    "450_points": 36,
    "600_points": 48
}

WdBorderType = {
    "border_top": -1,
    "border_left": -2,
    "border_bottom": -3,
    "border_right": -4,
    "border_horizontal": -5,
    "border_vertical": -6
}

WdLineStyle = {
    "dash_dot": 5,
    "dash_dot_dot": 6,
    "dash_dot_stroked": 20,
    "dash_large_gap": 4,
    "dash_small_gap": 3,
    "dot": 2,
    "double": 7,
    "double_wavy": 19,
    "emboss_3d": 21,
    "engrave_3d": 22,
    "inset": 24,
    "none": 0,
    "outset": 23,
    "single": 1,
    "single_wavy": 18,
    "thick_thin_large_gap": 16,
    "thick_thin_med_gap": 13,
    "thick_thin_small_gap": 10,
    "thin_thick_large_gap": 15,
    "thin_thick_med_gap": 12,
    "thin_thick_small_gap": 9,
    "thin_thick_thin_large_gap": 17,
    "thin_thick_thin_med_gap": 14,
    "thin_thick_thin_small_gap": 11,
    "triple": 8
}

WdColorIndex = {
    "black": 1,
    "blue": 2,
    "brightGreen": 4,
    "darkBlue": 9,
    "darkRed": 13,
    "darkYellow": 14,
    "gray25": 16,
    "gray50": 15,
    "green": 11,
    "pink": 5,
    "red": 6,
    "teal": 10,
    "turquoise": 3,
    "white": 8,
    "yellow": 7
}



if module == "new":
    try:
        ms_word = win32com.client.DispatchEx("Word.Application")
        word_document = ms_word.Documents.Add()
        ms_word.Visible = True
        mod_microsoft_word[session] = {
            "app": ms_word, 
            "doc": word_document
        }

    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e

if module == "open":
    path = GetParams("path")

    try:
        path = path.replace("/", os.sep)
        ms_word = win32com.client.DispatchEx("Word.Application")
        word_document = ms_word.Documents.Open(path)
        ms_word.Visible = True
        mod_microsoft_word[session] = {
            "app": ms_word, 
            "doc": word_document
        }
        
    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e

if module == "read":
    # Rocketbot params
    result = GetParams("result")
    details = GetParams("details")

    try:
        text = []
        word_document = mod_microsoft_word[session]["doc"]
        paragraphs = word_document.Paragraphs
        for paragraph in paragraphs:
            range_ = paragraph.Range
            font = range_.Font
            if details == "True":
                text.append({
                    "text": range_.Text,
                    "style": str(paragraph.Style),
                    "alignment": alignments(paragraph.Alignment),
                    "font": {
                        "name": font.Name,
                        "size": int(font.Size),
                        "bold": bool(font.Bold),
                        "italic": bool(font.Italic),
                        "underline": bool(font.Underline)
                    }
                })
            else:
                text.append("" + range_.Text)

        if result:
            SetVar(result, text)
    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e

if module == "addTable":
        
    rows = GetParams("rows")
    cols = GetParams("cols")
    tableStyle = GetParams("tableStyle")
    iframe = eval(GetParams("iframe"))
    lineType = iframe.get('lineType')
    lineSize = iframe.get('lineSize')
    
        
    try:

        word_document.Paragraphs.Add()
        paragraph = word_document.Paragraphs.Last
        range_ = paragraph.Range
        
        word_document.Tables.Add(Range=range_, NumRows=rows, NumColumns=cols)
        
        lastTable = word_document.Tables[str(word_document.Tables.count)]
        
        
        if tableStyle in WdBuiltinStyle:
            lastTable.Style = WdBuiltinStyle[tableStyle]
            
        try:
            for border in WdBorderType:
                lastTable.Borders(WdBorderType[border]).LineStyle = WdLineStyle[lineType]
                lastTable.Borders(WdBorderType[border]).LineWidth = WdLineWidth[lineSize]
        except Exception as e:
            print("\x1B[" + "31;40mError\x1B[" + "0m")
            print('El estilo de linea elegido no acepta ese tamano de linea. Por favor, escoja un tamano valido')
            PrintException()
            raise e
            



    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e
    
  
if module == "copyPasteText":
    
    startRange = GetParams("startRange")
    endRange = GetParams("endRange")
    path = GetParams("path")
    path = path.replace("/", os.sep)

    try:

        mod_microsoft_word[session] = {
            "app": ms_word, 
            "doc": word_document
        }
        
        word_document = mod_microsoft_word[session]["doc"]
        
        range_ = word_document.Range(Start=startRange, End=endRange)
        range_.Copy()
        
        

        word_documentPaste = ms_word.Documents.Open(path)
        ms_word.Visible = True

        word_documentPaste.Paragraphs.Last.Range.PasteAndFormat(Type=16)
        word_documentPaste.Paragraphs.Add()
        
        word_documentPaste.Save()
        word_documentPaste.Close()
        

    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e


if module == "copyText":
    
    startRange = GetParams("startRange")
    endRange = GetParams("endRange")
    
    try:
        mod_microsoft_word[session] = {
            "app": ms_word, 
            "doc": word_document
        }

        word_document = mod_microsoft_word[session]["doc"]
        
        range_ = word_document.Range(Start=startRange, End=endRange)
        range_.Copy()
    
    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e

if module == "pasteText":
    
    try:
        mod_microsoft_word[session] = {
            "app": ms_word, 
            "doc": word_document
        }
    
        word_document = mod_microsoft_word[session]["doc"]
        word_document.Paragraphs.Last.Range.PasteAndFormat(Type=16)
        word_document.Paragraphs.Add()

    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e
    
if module == "countCharacters":
    
    numParagraph = GetParams("numParagraph")
    result = GetParams("result")
    
    try:
        word_document = mod_microsoft_word[session].get("doc", None)
        count = word_document.Paragraphs.count
    
        paragraph =  word_document.Paragraphs(numParagraph)
        range_ = paragraph.Range
        characters = range_.Characters.Count - 1
        
        SetVar(result, characters)
        
    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e
    
    
    
if module == "readTable":
    try:

        result = GetParams("result")
        tableToRead = GetParams("tableToRead")
        
        
        tablesDoc = []

        if tableToRead:
            table_ = []
            table = word_document.tables(tableToRead)
            for row in table.rows:
                array_row = []
                for cell in row.cells:
                    """if len(array_row) > 0:
                        if array_row[-1] != cell.text:
                            array_row.append(cell.text)
                    else:
                        print("test")"""
                    information = cell.range.text
                    
                    text1 = re.sub(r"[\r\x07,]", "", information)
                    # text2 = re.sub(r"[\x07]", "", text1)
                    # text3 = re.sub(", , ", ", ", text2)
                    
                    array_row.append(text1)
                table_.append(array_row)
            if result:
                SetVar(result, table_)
        else:
            for table in word_document.tables:
                table_ = []
                for row in table.rows:
                    array_row = []
                    for cell in row.cells:
                        """if len(array_row) > 0:
                            if array_row[-1] != cell.text:
                                array_row.append(cell.text)
                        else:
                            print("test")"""
                        information = cell.range.text
                        
                        text1 = re.sub(r"[\r\x07,]", "", information)
                        # text2 = re.sub(r"[\x07]", "", text1)
                        # text3 = re.sub(", , ", ", ", text2)
                        
                        array_row.append(text1)
                    table_.append(array_row)
                tablesDoc.append(table_)
            if result:
                SetVar(result, tablesDoc)
            
        
    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e
    
    
if module == "editTable":
    
    numTable = GetParams("numTable")
    session = GetParams("session")
    deleteRow = GetParams("deleteRow")
    deleteColumn = GetParams("deleteColumn")
    addRow = GetParams("addRow")
    addColumn = GetParams("addColumn")
    columnWidth = GetParams("columnWidth")
    rowHeight = GetParams("rowHeight")
    
    try:
        
    
        table_ = word_document.Tables[numTable]

        if deleteRow:
            table_.Rows(deleteRow).Delete()
        if deleteColumn:
            table_.Columns(deleteColumn).Delete()
        if columnWidth:
            table_.Columns.SetWidth(columnWidth, 0)
        if rowHeight:
            table_.Rows.SetHeight(rowHeight, 2)
        if addRow == "True":
            table_.Rows.Add()
        if addColumn == "True":
            table_.Columns.Add()
    
    
    
    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e
    
    
    
    
    
    
if module == "deleteParagraph":
    
    try:
        numParagraph = GetParams("numParagraph")
        var = GetParams("variable")
        word_document = mod_microsoft_word[session].get("doc", None)
        
        count = word_document.Paragraphs.count
        
        
        if int(numParagraph) < count:
            paragraph =  word_document.Paragraphs(numParagraph)
            range_ = paragraph.Range
            SetVar(var, range_)
            range_.Delete()
        else:
            range_ = word_document.Paragraphs.Last.Range
            SetVar(var, range_)
            range_.Delete()
            
        
        
    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e
    
     

if module == "addTextBookmark":

    import copy

    bookmark_searched = GetParams("bookmark")
    text = GetParams("text")
    #clean = GetParams("Clean")
    #print(clean)

    try:
        """tmp_doc = Document()
        # Generate content in tmp_doc document
        tmp_doc.add_paragraph(text)
        # Reference the tmp_doc XML content
        tmp_doc_body = tmp_doc._element.body

        ele = document._element[0]
        bookmarks_list = ele.findall('.//' + qn('w:bookmarkStart'))
        for bookmark in bookmarks_list:
            # print(bookmark)
            name = bookmark.get(qn('w:name'))
            if name == bookmark_searched:
                par = bookmark.getparent()

                if clean:
                    next_element = bookmark.getnext()
                    if not isinstance(next_element, docx.oxml.CT_R):
                        next_element = next_element.getnext()
                    t = next_element.findall('.//' + qn('w:t'))
                    if len(t) == 1:
                        t[0].text = text
                elif isinstance(par, docx.oxml.CT_P):
                    bookmark_par_parent = par.getparent()
                    index = bookmark_par_parent.index(par)
                    for child in tmp_doc_body:
                        bookmark_par_parent.insert(index, child)
                        index = index + 1

                break
            else:
                name = None

        if not name:
            raise Exception("Bookmark not found")"""
        if word_document.Bookmarks.Exists(bookmark_searched):
            range = word_document.Bookmarks(bookmark_searched).Range
            range.Text = text
            #word_document.Bookmarks.Add(bookmark_searched)
        else:
            pass
    except Exception as e:
        PrintException()
        raise e

if module == "save":

    path = GetParams("path")
    try:
        if path:
            word_document.SaveAs2(path)
        else:
            word_document.SaveAs2()
    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e

if module == "to_pdf":
    to = GetParams("to")
    wdFormatPDF = 17
    try:

        word_document.ExportAsFixedFormat(OutputFileName=to, ExportFormat=wdFormatPDF, IncludeDocProps=True)

    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e

if module == "write":

    text = GetParams("text")
    type_ = GetParams("type")
    level = GetParams("level")
    align = GetParams("align")
    color = GetParams("color")
    size = GetParams("size")
    bold = GetParams("bold")
    italic = GetParams("italic")
    underline = GetParams("underline")

    try:
        word_document.Paragraphs.Add()
        paragraph = word_document.Paragraphs.Last
        range_ = paragraph.Range
        range_.Text = text
        font = paragraph.Range.Font

        size = float(size) if size else 12

        if color == None:
            color = "black"

        style = type_ + level if level is not None else ""


        if style in WdBuiltinStyle:
            paragraph.Style = WdBuiltinStyle[style]
        elif (type_ == "number" or type_ == "bullet") and int(level) > 5:
            level = 5
            style = type_ + str(level)
            paragraph.Style = WdBuiltinStyle[style]
        else:
            style = type_
            paragraph.Style = WdBuiltinStyle[style]
            
        font.ColorIndex = WdColorIndex[color]
        paragraph.Alignment = int(align) if align else 0
        font.Size = size
        
        if bold == "True":
            boldInt = -1
        else:
            boldInt = 0
        font.Bold = boldInt
        if italic == "True":
            italicInt = -1
        else:
            italicInt = 0
        font.Italic = italicInt
        if underline == "True":
            underlineInt = -1
        else:
            underlineInt = 0
        font.Underline = underlineInt
            
    except Exception as e:
        PrintException()
        raise e

if module == "close":

    try:
        ms_word = mod_microsoft_word[session]["app"]
        word_document = mod_microsoft_word[session]["doc"]
        word_document.Close()
        ms_word.Quit()
        del mod_microsoft_word[session]

    except Exception as e:
        PrintException()
        raise e

if module == "new_page":
    try:
        word_document.Paragraphs.Add()
        paragraph = word_document.Paragraphs.Last
        paragraph.Range.InsertBreak()
    except Exception as e:
        PrintException()
        raise e

if module == "add_pic":
    img_path = GetParams("img_path")

    try:
        # Only work with \
        img_path = img_path.replace("/", os.sep)

        count = word_document.Paragraphs.Count  # Count number paragraphs
        if count > 1:
            word_document.Paragraphs.Add()

        paragraph = word_document.Paragraphs.Last
        img = paragraph.Range.InlineShapes.AddPicture(FileName=img_path, LinkToFile=False, SaveWithDocument=True)
        print(img)
    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e
if module == "count_paragraphs":
    number = GetParams("variable")
    try:
        count = word_document.Paragraphs.count
        SetVar(number, count)

    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e

if module == "search_replace_text":
    text_search = GetParams("text_search")
    text_replace = GetParams("text_replace")
    numParagraphs = GetParams("numParagraphs")
    if text_search == text_replace:
        pass
    else:

        if numParagraphs:
            paragraphList = [int(s) for s in numParagraphs.split(',')]
            for i in paragraphList:
                paragraph = word_document.Paragraphs(i)
                range_ = paragraph.Range
                if text_search in range_.Text:
                    range_.Text = range_.Text.replace(text_search, text_replace)
        else:
            paragraphs = word_document.Paragraphs
            #fullRange = word_document.content
            for paragraph in paragraphs:
                range_ = paragraph.Range
                print(range_.Find.Text)
                range_.Find.Text = text_search
                range_.Find.Replacement.Text = text_replace
                range_.Find.Execute(Replace=2,Forward=True,MatchWholeWord=True)
                #print(range_.Find.Execute(FindText=text_search, ReplaceWith="text_replace", Replace=2))
                #if text_search in range_.Text:
                    #range_.Text = range_.Text.replace(text_search,text_replace)

if module == "search_text":
    try:
        text_search = GetParams("text_search")
        whichParagraph = GetParams("variable")
        paragraphList = []
        count = 1
        for paragraph in word_document.Paragraphs:
            range_ = paragraph.Range
            range_.Find.Text = text_search
            if range_.Find.Execute(Forward=True, MatchWholeWord=True):
                paragraphList.append(count)
            count += 1
        SetVar(whichParagraph, paragraphList)
        print(paragraphList)
    except Exception as e:
        print("\x1B[" + "31;40mError\x1B[" + "0m")
        PrintException()
        raise e
